### PR TITLE
Amend `release/1.8.3-oob`

### PR DESCRIPTION
**Pull request by johndebord**
##### Previous Behavior
When utilizing the `eosio.multisig` contract a user would perform the action
`propose` and specify (among the proposed transaction and other things) a number
of delay seconds in the transaction header. After proposing the transaction,
users would perform the action `approve` and after, the action `exec`; where if
the approvals of the proposal have been satisfied, the packed transaction (given
in the action `propose`) would get sent out as a deferred transaction that is
delayed by the specified number of delay seconds.

##### Current Behavior
The changes are two-fold:
1) The number of delay seconds is used to calculate the earliest possible
execution time of the action `exec` (of course, once the approvals of the
proposal have been satisfied). The earliest execution time is the point in time
at which the proposal has been satisfied plus the delayed seconds specified in
the transaction header of the packed transaction. Also note, that if a proposal
reverts its status of having the approvals satisified, then unsatisfied, and
then satisfied once again, the earliest execution time will always get
recalculated.
2) The given transaction (given in the action `propose`) is executed immediately
upon the user performing the action `exec` as a sequence of inline actions, but
only if the the proposal has been approved and the given number of delay seconds
has passed. Otherwise, the execution of the action `exec` shall fail.

### CODE REVIEW

File: contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp [hunk 1]
```diff
@@ -6,21 +6,235 @@
...
+          * */
+         template <typename U>
+         constexpr auto value_or( U&& def ) -> std::enable_if_t<std::is_convertible<U, T>::value, T&>& {
+            if (_has_value)
+               return _get();
+            return def;
+         }
+         constexpr T&& value_or()&& {
+            if (!_has_value)
+               return std::move(T());
```
Comment (tbfleming) [inline contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp:110]: This is a copy of the CDT's version, which has a serious bug here.

File: contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp [hunk 2]
```diff
@@ -123,52 +325,50 @@ namespace eosio {
...
+      time_point       time;
+   };
 
-            uint64_t primary_key()const { return proposal_name.value; }
-         };
-         typedef eosio::multi_index< "approvals2"_n, approvals_info > approvals;
+   struct [[eosio::table]] approvals_info {
+      uint8_t                 version = 1;
+      name                    proposal_name;
+      //requested approval doesn't need to cointain time, but we want requested approval
```
Comment (jgiszczak) [inline contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp:356]: "cointain" -> "contain"

File: contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp [hunk 3]
```diff
@@ -123,52 +325,50 @@ namespace eosio {
...
+      //to be of exact the same size ad provided approval, in this case approve/unapprove
```
Comment (jgiszczak) [inline contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp:357]: "to be of exact the same size ad provided" -> "to be of exactly the same size as provided"

File: contracts/eosio.msig/src/eosio.msig.cpp [hunk 4]
```diff
@@ -42,29 +72,38 @@ void multisig::propose( ignore<name> proposer,
...
+         for ( auto& level : requested ) {
+            a.requested_approvals.push_back( approval{ level, time_point{ microseconds{0} } } );
+         }
+      });
 }
 
 void multisig::approve( name proposer, name proposal_name, permission_level level,
                         const eosio::binary_extension<eosio::checksum256>& proposal_hash )
 {
+   // if ( level.permission == "eosio.code"_n ) {
```
Comment (jgiszczak) [inline contracts/eosio.msig/src/eosio.msig.cpp:95]: Second commit says it was removing commented code, but this is still present.  Are we keeping it as independent reference?

File: contracts/eosio.msig/src/eosio.msig.cpp [hunk 5]
```diff
@@ -90,9 +129,30 @@ void multisig::approve( name proposer, name proposal_name, permission_level leve
...
+            });
+         }
+      }
+   } else {
+      check( trx_header.delay_sec.value == 0, "old proposals are not allowed to have non-zero `delay_sec`; cancel and retry" );
+   }
 }
 
 void multisig::unapprove( name proposer, name proposal_name, permission_level level ) {
+   // if ( level.permission == "eosio.code"_n ) {
```
Comment (jgiszczak) [inline contracts/eosio.msig/src/eosio.msig.cpp:150]: Commented code.

File: contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp (hunk 3 above)
Comment (johndebord) [reply contracts/eosio.msig/include/eosio.msig/eosio.msig.hpp]: Updated comment.

File: contracts/eosio.msig/src/eosio.msig.cpp (hunk 4 above)
Comment (johndebord) [reply contracts/eosio.msig/src/eosio.msig.cpp]: Removed commented out code.

File: contracts/eosio.msig/src/eosio.msig.cpp (hunk 5 above)
Comment (johndebord) [reply contracts/eosio.msig/src/eosio.msig.cpp]: Remove commented out code.

**Review by jgiszczak — APPROVED** [review]

**Review by b1bart — CHANGES_REQUESTED** [review]
placeholder

**Review by b1bart — APPROVED** [review]

**Pull request merged**